io/src/main/scala/fs2/io/udp/AsynchronousSocketGroup.scala

```diff
@@ -0,0 +1,146 @@
+
+import java.nio.ByteBuffer
+import java.nio.channels.{DatagramChannel, Selector, SelectionKey, ClosedChannelException}
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.locks.ReentrantLock
+
+sealed trait AsynchronousSocketGroup {
+  private[udp] def register(channel: DatagramChannel): (AsynchronousSocketGroup.Attachment, SelectionKey)
+  private[udp] def enqueue(f: => Unit): Unit
+}
+
+object AsynchronousSocketGroup {
+
+  private[udp] class Attachment(
+    readers: MutableQueue[Either[Throwable,Packet] => Unit] = MutableQueue(),
+    writers: MutableQueue[(Packet,Option[Throwable] => Unit)] = MutableQueue()
+  ) {
+
+    def dequeueReader: Option[Either[Throwable,Packet] => Unit] = {
+      if (readers.isEmpty) None
+      else Some(readers.dequeue())
+    }
+
+    def queueReader(reader: Either[Throwable,Packet] => Unit): Unit = {
+      readers += reader
+      ()
+    }
+
+    def hasReaders: Boolean = readers.nonEmpty
+
+    def peekWriter: Option[(Packet,Option[Throwable] => Unit)] = {
+      if (writers.isEmpty) None
+      else Some(writers.head)
+    }
+
+    def dequeueWriter: Option[(Packet,Option[Throwable] => Unit)] = {
+      if (writers.isEmpty) None
+      else Some(writers.dequeue())
+    }
+
+    def queueWriter(writer: (Packet,Option[Throwable] => Unit)): Unit = {
+      writers += writer
+      ()
+    }
+
+    def hasWriters: Boolean = writers.nonEmpty
+
+    def close: Unit = {
+      readers.foreach { cb => cb(Left(new ClosedChannelException)) }
+      readers.clear
+      writers.foreach { case (_, cb) => cb(Some(new ClosedChannelException)) }
+      writers.clear
+    }
+  }
+
+  def apply(): AsynchronousSocketGroup = new AsynchronousSocketGroup {
+
+    private val selectorLock = new ReentrantLock()
+    private val selector = Selector.open()
+    private val pendingThunks: MutableQueue[() => Unit] = MutableQueue()
+
+    override def register(channel: DatagramChannel): (AsynchronousSocketGroup.Attachment, SelectionKey) = {
+      channel.configureBlocking(false)
+      val attachment = new AsynchronousSocketGroup.Attachment()
+      val key = {
+        selectorLock.lock
+        try {
+          selector.wakeup
+          channel.register(selector, 0, attachment)
+        } finally {
+          selectorLock.unlock
+        }
+      }
+      (attachment, key)
+    }
+
+    override def enqueue(f: => Unit): Unit = {
+      pendingThunks.synchronized {
+        pendingThunks.+=(() => f)
+      }
+      selector.wakeup()
+      ()
+    }
+
+    private val doneNow = new AtomicBoolean(false)
+    private val selectorThread = new Thread(new Runnable {
+      def run = try {
+        val readBuffer = ByteBuffer.allocate(1 << 16)
+        while (!doneNow.get && !Thread.currentThread.isInterrupted) {
+          pendingThunks.synchronized { while (pendingThunks.nonEmpty) pendingThunks.dequeue()() }
+          selectorLock.lock
+          selectorLock.unlock
```

Comment:
Yeah intended -- the selector lock is used as a barrier so that we don't select before we can register a new channel.